core/src/main/java/ai/timefold/solver/core/impl/score/stream/bavet/BavetConstraintFactory.java

```diff
@@ -92,20 +98,41 @@ public <Stream_ extends BavetAbstractConstraintStream<Solution_>> Stream_ share(
     // ************************************************************************
     // from
     // ************************************************************************
+    // Required for node sharing, since using a lambda will create different instances
+    private record ForEachFilteringCriteriaPredicateFunction<Solution_, A>(EntityDescriptor<Solution_> entityDescriptor,
+            ForEachFilteringCriteria criteria) implements Function<ConstraintNodeBuildHelper<Solution_, ?>, Predicate<A>> {
+        public Predicate<A> apply(@NonNull ConstraintNodeBuildHelper<Solution_, ?> helper) {
+            return helper.getForEachPredicateForEntityDescriptorAndCriteria(entityDescriptor, criteria);
+        }
+
+        @Override
+        public boolean equals(Object object) {
+            if (object instanceof ForEachFilteringCriteriaPredicateFunction<?, ?> that) {
+                return criteria == that.criteria
+                        && Objects.equals(entityDescriptor.getEntityClass(), that.entityDescriptor.getEntityClass());
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(entityDescriptor.getEntityClass(), criteria);
+        }
```

Comment:
As the record doesn't use any arrays, reimplementing equals/hashCode should not be necessary. This is not performance-sensitive code, so micro-optimizations that may be possible in custom impls do not apply here either.

If the goal is to compare entity descriptors by class, then I suggest we add equals to entity descriptor; that is where this check logically belongs. Two entity descriptors for the same class are a practical impossibility; they _should_ equal.